components/builder-depot/src/data_store.rs

```diff
@@ -327,56 +329,152 @@ impl Bucket for PackagesIndex {
 /// This is how packages will be "promoted" between environments without duplicating data on disk.
 pub struct ChannelsTable {
     pool: Arc<ConnectionPool>,
+    channel_package_map: HashMap<String, Vec<depotsrv::PackageIdent>>,
+    origin_channel_map: HashMap<String, Vec<String>>,
     pub pkg_channel_idx: PkgChannelIndex,
     pub channel_pkg_idx: ChannelPkgIndex,
 }
 
+// It's worth noting that these two new HashMaps are a temporary measure, put in place to avoid
+// having to deal with persisting things to Redis while the work to convert our persistence layer
+// to Postgres is in progress. Once the transition to PG is complete, these two HashMaps
+// should go away and all of the methods below that use them should be changed to use PG instead.
+
 impl ChannelsTable {
     pub fn new(pool: Arc<ConnectionPool>) -> Self {
         let pool1 = pool.clone();
         let pool2 = pool.clone();
         let pkg_channel_idx = PkgChannelIndex::new(pool1);
         let channel_pkg_idx = ChannelPkgIndex::new(pool2);
+        let channel_package_map = HashMap::<String, Vec<depotsrv::PackageIdent>>::new();
+        let origin_channel_map = HashMap::<String, Vec<String>>::new();
 
         ChannelsTable {
             pool: pool,
+            channel_package_map: channel_package_map,
+            origin_channel_map: origin_channel_map,
             pkg_channel_idx: pkg_channel_idx,
             channel_pkg_idx: channel_pkg_idx,
         }
     }
 
-    pub fn all(&self) -> Result<Vec<String>> {
-        let conn = self.pool.get().unwrap();
-        match conn.smembers(Self::prefix()) {
-            Ok(members) => Ok(members),
-            Err(e) => Err(Error::from(e)),
-        }
+    pub fn all(&mut self, origin: &str) -> Vec<String> {
+        let channels = self.origin_channel_map
+            .entry(origin.to_string())
+            .or_insert(vec!["stable".to_string(), "unstable".to_string()]);
+        channels.clone()
     }
 
-    pub fn associate(&self, channel: &str, pkg: &depotsrv::Package) -> Result<()> {
-        let script = redis::Script::new(r"
-            redis.call('sadd', KEYS[1], ARGV[2]);
-            redis.call('zadd', KEYS[2], 0, ARGV[1]);
-        ");
-        try!(script.arg(pkg.get_ident().to_string())
-            .arg(channel.clone())
-            .key(PkgChannelIndex::key(&pkg.get_ident()))
-            .key(ChannelPkgIndex::key(&channel.to_string()))
-            .invoke(self.pool.get().unwrap().deref()));
+    pub fn create(&mut self, origin: String, channel: String) -> Result<()> {
+        let mut vec = match self.origin_channel_map.get(&origin) {
+            Some(channels) => channels.clone(),
+            None => Vec::new(),
+        };
+
+        if self.channel_exists(&origin, &channel) {
+            return Err(Error::Generic("Channel already exists".to_string()));
```

Comment:
The part of me that loves idempotent API's looks at errors like this one and wonders if its really an error, at least when it reaches the end user. I don't have a strong opinion either way, but I think my preference would be to lean towards the APIs being idempotent, and just returning true (at the top of the stack - this error might well remain, because you need to early return anyway, and we can always change our mind.)